app/scripts/directives/serviceIntegration.js

```diff
@@ -0,0 +1,242 @@
+          // update the binding with consumer and provider metadata annotations
+          DataService.update(bindingPreferredVersion, watchBinding.metadata.name, watchBinding, context)
+          // then retrieve the deployment
+          .then(function() {
+            return DataService.get(
+              deploymentPreferredVersion, 
+              _.get(ctrl, 'consumerService.metadata.labels.serviceName'), 
+              context, {errorNotification: false}
+            );
+          })
+          // then add the enabled service metadata label
+          .then(function(deployment) {
+            deployment.spec.template.metadata.labels[_.get(ctrl.serviceInstance, 'metadata.labels.serviceName')] = "enabled";
+            // and update the deployment and trigger a redeploy
+            return DataService.update(
+              deploymentPreferredVersion, 
+              _.get(ctrl, 'consumerService.metadata.labels.serviceName'), 
+              deployment, 
+              context
+            );
+          })
+          .catch(function(err) {
+            NotificationsService.addNotification({
+              type: "error",
+              message: "Failed to integrate service binding",
+              details: err.data.message
+            });
+          });
+        }
+      });
+      
+      //create the pod preset asynchronously to the binding
+      DataService.create(version, null, podPreset, context)
+      .catch(function(err) {
+        NotificationsService.addNotification({
+          type: "error",
+          message: "Failed to create pod preset",
+          details: err.data.message
+        });
+      });
+      
+    };
+
+    ctrl.getState = function(){
+      var statePending = 2;
+      var stateActive = 1;
+      var stateNoBinding = 0;
+      var stateNoService = -1;
+
+      if(ctrl.podPreset && !ctrl.binding) {
+        return statePending;
+      }
+      if(ctrl.podPreset && ctrl.binding) {
+        //pod preset and binding exist, state 1
+        return stateActive;
+      }
+      if(ctrl.binding) {
+        if(isBindingReady(ctrl.binding)){
+          return stateActive;
+        }
+        return statePending;
+      }
+      if (ctrl.serviceInstance){
+        return stateNoBinding;
+      }
+      return stateNoService;
+    };
+
+    //called in callback from succesful delete-link for binding
+    ctrl.deletePodPreset = function(){
+      var context = {namespace: ctrl.consumerService.metadata.namespace};
+      var deleteOptions = {propagationPolicy: null};
+      DataService.delete(podPresetPreferredVersion, ctrl.podPreset.metadata.name, context, deleteOptions)
+      .then(function(){
+        return DataService.get(
+          deploymentPreferredVersion, 
+          _.get(ctrl, 'consumerService.metadata.labels.serviceName'), 
+          context
+        );
+      })
+      .then(function(deployment) {
+        delete deployment.spec.template.metadata.labels[integrationName];
```

Comment:
Use `angular.copy` before changing the deployment. DataService does some caching and you don't want to modify the cached data.